[ZEPPELIN-5442] Allow users to set a separate namespace for the interpreter

rickchengx opened the pull request:
### What is this PR for?
Currently, if zeppelin is running under k8s mode and the zeppelin server is running inside the k8s cluster, it will start the interpreter pod under the namespace specified in `/var/run/secrets/kubernetes.io/serviceaccount/namespace`. But if zeppelin-server is running **outside of kubernetes cluster**, it will use the config `zeppelin.k8s.namespace` to start the interpreter pod under the given namespace. However, this approach will cause users to only set a **global namespace**.

Hence, this PR allows user to set a set **separate namespace** for the interpreter in the interpreter settings:

Example usage:
```
%spark.conf
zeppelin.k8s.interpreter.namespace spark
```
The spark interpreter pod will be created under the namespace: `spark`.

```
%python.conf
zeppelin.k8s.interpreter.namespace python
```
The python interpreter pod will be created under the namespace: `python`.

### What type of PR is it?
[Improvement]

### Todos
* [ ] - Task

### What is the Jira issue?
* <https://issues.apache.org/jira/browse/ZEPPELIN-5442>

### How should this be tested?
* CI pass and manually tested

### Screenshots (if appropriate)

### Questions:
* Does the licenses files need update? No
* Is there breaking changes for older versions? No
* Does this needs documentation? No

Reamer commented on zeppelin-plugins/launcher/k8s-standard/src/main/java/org/apache/zeppelin/interpreter/launcher/K8sStandardInterpreterLauncher.java:
```diff
@@ -73,6 +73,11 @@ String getNamespace() throws IOException {
     if (isRunningOnKubernetes()) {
       return readFile(Config.KUBERNETES_NAMESPACE_PATH, Charset.defaultCharset()).trim();
     } else {
+      //namespace may be set from the interpreter properties
```
This feature should also be available with a Zeppelin server running on Kubernetes.

Reamer reviewed (changes requested):
Feature should also be available for Zeppelin server running in K8s + CI failures.

rickchengx commented:
> Feature should also be available for Zeppelin server running in K8s + CI failures.

Hi, @Reamer 
Thanks for your comment. I have fixed the CI failure.
After my test, when zeppelin server is also running inside k8s, there'll be a few more problems:
1. the zeppelin server needs to obtain the **cluster role** instead of role to create pods and services in other namespaces.
2. There're 2 namespaces: the namespace of zeppelin server and the namespace of zeppelin interpreter. So the namespace return by method `getNamespace` may be wrong if we return the `zeppelin.k8s.interpreter.namespace` while zeppelin server is also running in k8s.
3. When zeppelin server pod and interpreter pod are in different namespaces, the `ownerReference` can not be set. Since Cross-namespace owner references are disallowed by design <https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/>.

But if zeppelin server is running outside k8s, it can correctly create the interpreter pod in different namespaces. There won't be a problem of cross-namespace communication between zeppelin server and interpreter.

So how about support the feature of separate namespace while zeppelin server is running in k8s in another PR.

Reamer commented on zeppelin-plugins/launcher/k8s-standard/src/main/java/org/apache/zeppelin/interpreter/launcher/K8sRemoteInterpreterProcess.java:
```diff
@@ -275,7 +275,7 @@ Properties getTemplateBindings(String userName) {
 
     // k8s template properties
-    k8sProperties.put("zeppelin.k8s.namespace", getNamespace());
+    k8sProperties.put("zeppelin.k8s.interpreter.namespace", getNamespace());
```
This line is a breaking change. Please correct your PR description. People need to correct their own `interpreter-spec.yaml`.

Reamer replied:
I think this should be the place where you need to call another method, something like `getInterpreterNamespace()`, where the default value is the output of `getServerNamespace()`.

Reamer reviewed (changes requested):
Thank you very much for your analysis. I understand the problems when Zeppelin server is running in K8s and they are not easy to solve.
Please add the documentation which explain the flag. Please make it clear that this feature only works when Zeppelin is running outside of K8s.

rickchengx commented:
> Thank you very much for your analysis. I understand the problems when Zeppelin server is running in K8s and they are not easy to solve.
> Please add the documentation which explain the flag. Please make it clear that this feature only works when Zeppelin is running outside of K8s.

Hi, @Reamer 
Thanks for the comment. Due to potential code conflicts, I think it's better to achieve the feature of setting the separate namespace for the interpereter when Zeppelin server is running inside k8s. 

So I added several commits to address the above three problems. Specifically,
1. Change `role` and `rolebinding` to `clusterrole` and `clusterrolebinding`
2. Update the conditions for setting `ownerReference`
3. Return the correct namespace of Zeppelin Server


Hence, it can set the separate namespace for each interpreter no matter Zeppelin server is running inside k8s or not.

For example, user can start a spark interpreter in the namespace of `spark`:
<img width="1403" alt="截屏2021-07-21 下午5 05 45" src="https://user-images.githubusercontent.com/38122586/126467521-975c35b0-d0fa-43c3-b7d9-7ef548752d6e.png">
<img width="866" alt="截屏2021-07-21 下午5 39 38" src="https://user-images.githubusercontent.com/38122586/126467929-6c7fc835-1e68-4d5a-ac95-f6483f5b4bb3.png">

Meanwhile, user can start a shell interpreter in the namespace of `sh`:
<img width="1419" alt="截屏2021-07-21 下午5 02 55" src="https://user-images.githubusercontent.com/38122586/126468069-162d588a-f43a-4f01-9cd3-1d0aa35bdc37.png">
<img width="345" alt="截屏2021-07-21 下午5 03 04" src="https://user-images.githubusercontent.com/38122586/126468113-35c1214d-dfa4-4514-854a-6b5e888f63ec.png">

Reamer commented on k8s/zeppelin-server.yaml:
```diff
@@ -213,14 +213,15 @@ rules:
 subjects:
 - kind: ServiceAccount
   name: zeppelin-server
+  namespace: default
```
It is not sure that the `zeppelin-server` ServiceAccount is part of the `default` namespace.

rickchengx replied:
Yes, the namespace of zeppelin server is not sure. But the `clusterrolebinding` needs to specify the namespace of the service account, which is applied by user when user is creating the zeppelin server in k8s.

So if user does not apply the `zeppelin-server` in the default namespace, he also needs to modify this `namespace` field in the `clusterrolebinding`.

Reamer replied:
Instead of creating a `ClusterRole` and a `ClusterRoleBinding`, I would prefer an additional blueprint with a `Role` and a `RoleBinding`. Inside the `RoleBinding` the user has to replace the `Namespace`.

Reamer commented on zeppelin-plugins/launcher/k8s-standard/src/main/java/org/apache/zeppelin/interpreter/launcher/K8sStandardInterpreterLauncher.java:
```diff
@@ -70,6 +70,11 @@ boolean isRunningOnKubernetes() {
    * @throws IOException if namespace file could not be read
    */
   String getNamespace() throws IOException {
+    //namespace may be set from the interpreter properties
```
The `K8sStandardInterpreterLauncher` is executed only on the Zeppelin server side.
Please rename this function to `getInterpreterNamespace()`.

Reamer commented on zeppelin-plugins/launcher/k8s-standard/src/main/java/org/apache/zeppelin/interpreter/launcher/K8sRemoteInterpreterProcess.java:
```diff
@@ -135,6 +139,43 @@ public String getNamespace() {
     return namespace;
```
Rename this variable to `serverNamespace` and change the function name to `getServerNamespace()`.

Reamer commented on zeppelin-plugins/launcher/k8s-standard/src/main/java/org/apache/zeppelin/interpreter/launcher/K8sStandardInterpreterLauncher.java:
```diff
@@ -95,9 +100,11 @@ String getHostname() {
   private String getZeppelinService(InterpreterLaunchContext context) throws IOException {
     if (isRunningOnKubernetes()) {
+      //The namespace of zeppelin server can only be read from Config.KUBERNETES_NAMESPACE_PATH while it runs in k8s cluster, it may be different from the namespace of interpreter
+      String serverNamespace = readFile(Config.KUBERNETES_NAMESPACE_PATH, Charset.defaultCharset()).trim();
```
Please move this part to a separate function. Something like `getServerNamespace()`.

Reamer reviewed (commented):
Some comments to improve readability.
In most cases it is not clear which namespace (server, interpreter) is used.
We should also avoid duplicate code. I think it makes sense to outsource code to the utility class.

rickchengx commented:
Hi, @Reamer 
Thanks for the comments. I have moved the common functions (`isRunningOnKubernetes()`, `readFile()`, `getInterpreterNamespace()` and `getCurrentK8sNamespace()`) into `org.apache.zeppelin.interpreter.launcher.K8sUtils` to avoid duplicate code.

And I also make it more clear to distinguish the namespace of server and interpreter.

Reamer commented on zeppelin-plugins/launcher/k8s-standard/src/main/java/org/apache/zeppelin/interpreter/launcher/K8sUtils.java:
```diff
@@ -74,4 +82,61 @@ private static long convertToBytes(String memory) {
+    return new File(Config.KUBERNETES_NAMESPACE_PATH).exists();
+  }
+
+  public static String readFile(String path, Charset encoding) throws IOException {
```
Please change this to `private`.

rickchengx replied:
Sure.

rickchengx commented:
> Instead of creating a ClusterRole and a ClusterRoleBinding, I would prefer an additional blueprint with a Role and a RoleBinding. Inside the RoleBinding the user has to replace the Namespace.

But if `zeppelin server` needs to create interpreter pod in another namespace, its service account needs to be granted a `clusterrolebinding` because a `rolebinding` only grants permissions within a specific namespace. What do you think?

Reamer commented:
In my opinion an additional step should be needed to give the Zeppelin server more rights on a namespace via a RoleBinding. Creating a ClusterRoleBinding gives too many rights to the zeppelin-server. With a ClusterRoleBinding the Zeppelin server has rights in namespaces that are not used by the Zeppelin server at all.

rickchengx commented:
> In my opinion an additional step should be needed to give the Zeppelin server more rights on a namespace via a RoleBinding. Creating a ClusterRoleBinding gives too many rights to the zeppelin-server. With a ClusterRoleBinding the Zeppelin server has rights in namespaces that are not used by the Zeppelin server at all.

Sorry for the late response. I agree that `clusterrole` may gives too many rights to the zeppelin server. But the zeppelin server  pod uses the service account created in <https://github.com/apache/zeppelin/blob/master/k8s/zeppelin-server.yaml#L199>, which is a namespaced resource.

Assuming that the zeppelin server and its service account are created in `default` namespace. And if the `zeppelin-server.yaml` uses the `rolebinding` to obtain the permissions in another namespace (such as `spark`), it cannot find the created service account in `default` namespace.

Or is there other ways to set the separate namespace with `role` and `rolebinding`?

Reamer commented:
To avoid creating the same `Role` in every namespace, a `ClusterRole` would be good.
You can then create a `RoleBinding` in each namespace, which has the Zeppelin server `ServiceAccount` as subject.
Example `RoleBinding` in `spark` namespace.

```
kind: RoleBinding
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: zeppelin-server-role-binding
  namespace: spark
subjects:
  - kind: ServiceAccount
    name: zeppelin-server
    namespace: default
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: zeppelin-server-role
```

I think that should be sufficient. This `RoleBinding` should be created manuell by the user for each namespace, which should be used be zepperlin.

rickchengx commented:
> To avoid creating the same `Role` in every namespace, a `ClusterRole` would be good.
> You can then create a `RoleBinding` in each namespace, which has the Zeppelin server `ServiceAccount` as subject.
> Example `RoleBinding` in `spark` namespace.
> 
> ```
> kind: RoleBinding
> apiVersion: rbac.authorization.k8s.io/v1
> metadata:
>   name: zeppelin-server-role-binding
>   namespace: spark
> subjects:
>   - kind: ServiceAccount
>     name: zeppelin-server
>     namespace: default
> roleRef:
>   apiGroup: rbac.authorization.k8s.io
>   kind: ClusterRole
>   name: zeppelin-server-role
> ```
> 
> I think that should be sufficient. This `RoleBinding` should be created manuell by the user for each namespace, which should be used be zepperlin.

Thanks for the comments. I have changed the `clusterrolebinding` to `rolebinding` and updated the docs. Now by default, user can create the interpreter pod in the `default` namespace. If the user wants to create and run the interpreter in another namespace, he also needs to add a `rolebinding` in the `zeppelin-server.yaml`

Reamer commented on zeppelin-plugins/launcher/k8s-standard/src/main/java/org/apache/zeppelin/interpreter/launcher/K8sRemoteInterpreterProcess.java:
```diff
@@ -287,10 +288,15 @@ Properties getTemplateBindings(String userName) {
+    String serverNamespace = K8sUtils.getCurrentK8sNamespace();
+    String interpreterNamespace = getInterpreterNamespace();
+    //Set the owner reference (zeppelin-server pod) for garbage collection when zeppelin server and the zeppelin interpreter is in the same namespace (Kubernetes cannot specify an owner in different namespace).
+    if (ownerUID() != null && ownerName() != null && serverNamespace != null && serverNamespace.equals(interpreterNamespace)) {
```
`K8sUtils.getCurrentK8sNamespace()` can return `null`. Please use [`StringUtils.equals`](https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/StringUtils.html#equals-java.lang.CharSequence-java.lang.CharSequence-)

Reamer commented:
> Now by default, user can create the interpreter pod in the `default` namespace.

Is it necessary to add a namespace?

Reamer commented on k8s/zeppelin-server.yaml:
```diff
@@ -217,10 +217,12 @@ kind: RoleBinding
 subjects:
 - kind: ServiceAccount
   name: zeppelin-server
+  namespace: default
```
Please remove the namespace parameter. Afterwards, the RoleBinding is created in the namespace in which the other resources are also created.

rickchengx replied:
Sure.

Reamer reviewed (changes requested):
Sorry for my late reply. Just a small change request.

rickchengx commented:
> Please remove the namespace parameter. Afterwards, the RoleBinding is created in the namespace in which the other resources are also created.

Sorry for the late response since I am busy with school stuff.
I have removed the namespace parameter and rebased this PR to avoid conflicts.

Reamer reviewed (approved):
LGTM
This PR introduces a breaking change. Please update your custom interpreter-spec.yaml.
`zeppelin.k8s.namespace` -> `zeppelin.k8s.interpreter.namespace`
